### PR TITLE
[trainers,hparams] feat: DMD2, TDM and TDM-R1 distillation on the multi-role runtime

### PR DESCRIPTION
**Pull request by Jayce-Ping**
Depends on #209 and remains stacked on `feature/structured-trajectory-runtime`.

## Summary

Adds production DMD2, TDM, and TDM-R1 trainers on the multi-role runtime, including deterministic multi-step replay, official TDM conditional re-noising and importance weighting, role-aware save/resume, and distributed backend correctness.

## Refactor and correctness

- Extracts shared distillation epoch, GAS, metric, role-phase, and TDM trajectory runtimes.
- Makes TDM-R1 inherit TDM directly while retaining its reward and surrogate objectives.
- Keeps DMD2 ODE replay bit-exact and rank-seeded.
- Implements official TDM conditional re-noising and density-ratio weighting.
- Preserves the TDM-R1 distribution anchor, teacher guidance, and learned-surrogate directions.

## Distributed behavior

- ZeRO-2 steps only the active role's optimizer groups.
- FSDP2 registry, EMA, and reference state bind to live DTensor parameters after prepare.
- FSDP1 clips through the prepared root and realigns resumed role dtypes before flattening.
- PEFT routing is restored after reference contexts and across role windows.
- TDM-R1 disables activation checkpointing on FSDP1 where interleaved reference forwards cannot be recomputed as the same graph; FSDP2 keeps it enabled.

## Verification

- [x] DMD suite: **1263 passed, 14 skipped**.
- [x] Final review uses bundled `dataset/ocr` directly, removes generated metadata and private launch details, and passes **110 targeted tests**.
- [x] Published validated SD3.5 OCR recipes for DMD2, TDM, and official-aligned TDM-R1.
- [x] TDM-R1 uses the released rank-32 initializer, G24 / 48 prompts, `beta_dpo=10`, and one fake, surrogate, and generator update per epoch.
- [x] TDM-R1 completed all three role updates under FSDP2; the production LoRA recipe uses DeepSpeed ZeRO-2.
- [x] Fixed CUDA endpoint rounding where a timestep below 1000 could map to sigma one; added CUDA and TDM boundary regressions.
- [x] **40/40** SD3.5 combinations passed on 16 GPUs: GRPO, NFT, DMD2, TDM, and TDM-R1 across DDP, ZeRO-2, FSDP1, and FSDP2 with LoRA and full finetuning.
- [x] Every combination completed backward, optimizer step, model-only checkpoint save, and resume.

## Stack

Merge order: **#209 → #210 → #212**.

- Parent: #209 (`98b6b4d`)
- This head: `e499a1c`
- Child: #212

### CODE REVIEW

**Comment by Jayce-Ping** [comment]
## Evaluation now runs, and the guided real score gets its negatives

Two follow-ups from a 64-GPU run where `eval_freq` fired on schedule and produced
nothing.

`DMD2` and `TDM` zeroed the whole feedback runtime to express that their loss is
reward-free. That contract is enforced in `Arguments`, which already rejects any training
reward for these trainers, so the override bought nothing and took the eval side down with
it: every eval dataset was skipped for want of a reward buffer. They now reuse the shared
runtime and assert only what is theirs, that no training reward survived.

Separately, preprocessing decides whether to encode negative prompts from one number and
it read `guidance_scale`. The configuration this algorithm wants -- a CFG-free generator
beside a guided real score -- puts 1.0 there, so negatives were never encoded and the
guided query fell back to unguided with a warning nobody was watching.
`get_preprocess_guidance_scale` now reports the largest scale any stage may ask for, which
is the same fix DGPO's `kl_cfg` already carries.

## Verification

- [x] 1146 tests pass
- [x] 16-GPU Flux2-Klein-9B DMD2, eval on: `Eval/text_rendering 4/4` completes and logs
      `eval/text_rendering/reward_ocr_mean=0.7150`
- [x] Checkpoint written at `save_freq=1` carries generator at the root and `roles/fake/`
- [x] Resuming that checkpoint loads `role=generator` and `role=fake` into their own
      variants, with no "roles kept their initial weights" warning
- [x] Three-role (TDM-R1 shape) save/load round trip added on the base branch, asserting
      no two roles collapsed onto the same weights
- [x] 64-GPU long run restarted on the standard configuration: 4-step generator,
      `real_guidance_scale=4.0`, fake:generator = 5:1, `eval_freq=10`

**Comment by Jayce-Ping** [comment]
## DMD2 backend matrix: DDP and ZeRO-2 are green, FSDP2 has two distinct gaps

Verified on SD3.5-medium, 16 GPUs across 2 nodes. Each entry trains one epoch, writes a
checkpoint, and a paired run resumes from it; both halves must pass.

| finetune | DDP | ZeRO-2 | FSDP2 |
| --- | --- | --- | --- |
| LoRA | train + resume | train + resume | fake role has no trainable parameters |
| full | train + resume | train + resume | reference snapshot is not a DTensor |

Three defects fixed to get there, each caught by the matrix rather than by unit tests:

**Role gradients were asserted where the backend does not leave them.** The
per-microbatch assertions read `parameter.grad`; DeepSpeed reduces into the engine's own
partitioned buffers and clears it, so the read reported "no gradients" for every role of
every correct run and rejected ZeRO-2 before its first step. The assertions now apply only
where `grad` is authoritative, and where they do apply they report what they inspected.

**`ref_param_device` was declared by six algorithms and read by all of them.**
`BaseAdapter._init_ref_parameters` reads it unconditionally, so every algorithm that did
not declare it died with an AttributeError the moment a full-weight run needed a
reference. Moved to the base arguments.

**A previous run's files outlived the checkpoint.** Re-running under the same `run_name`
left a shard index from a larger model beside a fresh single-file save; the loader trusts
the index and went looking for shards nobody wrote. Save now clears an entry's own
artifacts, leaving nested roles and components alone.

## The two FSDP2 gaps, stated precisely

The enriched assertion message identifies the first one outright:

```
active role 'fake' expected at least one gradient at the sync boundary, received none
across 382 parameter(s) of which 0 require grad; backend is DistributedType.FSDP
```

That is not gradient visibility. Under FSDP2 with LoRA the fake role owns the whole
component's 382 parameters and none of them require grad, so the role genuinely trains
nothing -- the same class of silent no-op as FSDP1 + LoRA, now on the second role. Worth
noting that the setup-time trainability guard cannot catch it: it asks whether *any* rank
holds *any* trainable parameter, which the generator satisfies on its own. A multi-role
run needs that guard per role.

The second is in the reference snapshot. DMD2 sets `requires_ref_model = True` because the
frozen pretrained score *is* its teacher; under LoRA that costs nothing, since disabling
the adapter recovers the base weights, but a full-weight run has to keep a copy. That copy
is stored as a plain tensor while the live parameter is a DTensor, so writing it back
raises `aten.copy_.default got mixed torch.Tensor and DTensor`. The snapshot has to be
taken in the same form as the parameter it will be written back into.

**Comment by Jayce-Ping** [comment]
## TDM now runs outside LTX2, and DDP + ZeRO-2 are green for both algorithms

| algorithm | finetune | DDP | ZeRO-2 |
| --- | --- | --- | --- |
| DMD2 | LoRA | train + resume | train + resume |
| DMD2 | full | train + resume | train + resume |
| TDM | LoRA | train + resume | train + resume |
| TDM | full | train + resume | train + resume |

TDM had never run on anything but LTX2. It checked its rollout precondition by reaching
into `sample.trajectory` and demanding a `StructuredTrajectory`, which only the LTX2 pair
publishes; every other adapter stores the same boundaries in the legacy layout. So TDM and
TDM-R1 failed identically on SD3.5 and Flux2-Klein, on every backend, before the first
step. The replay itself never had that restriction -- it reads through
`adapter.get_replay_step`, which understands both layouts -- so the precondition is now
expressed through the same accessor, with the structural checks kept for adapters that can
answer them.

Behind it sat a smaller one that surfaced three times in a row: the terminal transition's
next coordinate is shared by the whole batch and synthesized as an integer zero, so it
arrives as a scalar `int64` where every consumer expects a real `(B,)` coordinate. Fixing
it at each reader just moved the failure to the next one; the transition is now normalized
once, where TDM receives it.

## What is left

`aten.copy_.default got mixed torch.Tensor and DTensor` under FSDP2 with full weights: DMD2
keeps a copy of the pretrained score as its teacher, stored as a plain tensor while the
live parameter is a DTensor.

FSDP2 with LoRA reports `382 parameter(s) of which 0 require grad` for the fake role, so
the role trains nothing -- the same class of silent no-op as FSDP1 + LoRA, on the second
role. The setup-time trainability guard cannot catch it because it asks whether *any* role
has trainable parameters, which the generator satisfies alone.

**Comment by Jayce-Ping** [comment]
## Overnight update

Pushed 9 commits (`ab9f143..d3e6b48`).

### What landed

**Observability.** Distillation runs had no loss, no gradient norm, nothing to separate a diverging generator from a converging one until the images came back wrong. Per-role losses and gradient norms now flow for all three algorithms, plus the clean-latent statistics that catch a generator drifting out of the VAE range while its loss still falls.

**Sampler generalization.** TDM-R1 forced `group_contiguous` and therefore demanded `per_device_batch_size % group_size == 0`, which rejects the ordinary shape of one reward group spanning a global step. Whole groups per microbatch is a property of the sampler, not a constraint on the batch shape: `group_distributed` deals every rank an equal share of every group and sums the group logits across ranks, which `group_preference.py` already implemented and nothing called. The resolver now picks between the two.

**TDM-R1 objective aligned with the paper.** The generator mixed its reward in as `tdm_loss + tdm_weight * preference_loss`, while TDM-R1 follows three detached directions in clean-prediction space and mixes only the two rewards, leaving the distribution-matching anchor uncoefficiented. Under the old reading the published `tdm_weight=0.3` put the matching term an order of magnitude off. Also adds the surrogate's frozen-reference penalty, the trajectory-position weighting, and the PPO trust region against a lagging surrogate copy.

**Checkpoint conversion.** `scripts/convert_tdm_checkpoint.py` turns the official TDM / TDM-R1 releases (a positional tensor list with no names) into an adapter directory `resume_path` reads. Verified end to end on `Luo-Yihong/TDM_sd3-5_lora`.

### Bugs found along the way

- The distillation rollout helper never cleared the reward buffer, so from the second iteration of an epoch onward it scored all samples generated so far against only the current batch. Surfaced as a shape error deep in the advantage processor with nothing pointing at the rollout.
- The trust-region ratio compared two scores taken at independently sampled perturbations, so it measured the draw rather than the step taken: on the first epoch, where nothing should be clipped, half the batch was.
- Accelerate maps every reduction other than `"max"` onto a sum, which made a rank-agreement check compare a 16-rank sum against a maximum.

### Verification

1193 unit tests pass. An aligned TDM-R1 recipe (`general_ocr_sd3_8gpu_G24_4step`: 48x24 groups, 12 per device, 1:1:1 role updates, `tdm_weight=0.3`, `beta_dpo=10`, rollout CFG-free with `rl_cfg=4.5`) trains clean on 8 GPUs from the converted TDM LoRA.

### Known issue: FSDP2

FSDP2 is the one backend still failing, and the diagnostic added here names it exactly:

```
active role 'fake' expected at least one gradient at the sync boundary, received none.
Backend DistributedType.FSDP;
generator: 382 params, 382 require grad, 0 have grad;
fake:      382 params,   0 require grad, 0 have grad
```

The generator's parameters survive `prepare()` with `requires_grad` intact; the fake role's do not. The variant registry's recorded parameter objects appear to go stale when FSDP2 replaces them with sharded DTensors, so re-marking trainability writes to tensors the module no longer uses. DDP and DeepSpeed ZeRO-2 are unaffected. Not fixed here.

**Comment by Jayce-Ping** [comment]
## Replay/state root cause fixed

Pushed `a0f17ed` and `9cb97fc`.

The two-epoch replay failure was not checkpointing and not a tolerance issue. ULP diagnostics showed 16–512 fp16 ULP discrepancies. A pre-fake replay passed, and parameter probes then proved one fake ZeRO-2 step mutated 190/382 generator LoRA tensors (`max_abs=5.75e-06`) despite public generator grads being `None`.

Two fixes were required:

1. **ZeRO-2 role isolation.** Accelerate's DeepSpeed wrapper steps inside `backward()` and ZeRO iterates every internal `bit16_group`; native `grad is None` semantics do not apply. The role coordinator now filters DeepSpeed's per-group optimizer seam during backward, preserving inactive params, Adam moments, and step counters while leaving reduction/cleanup/all-gather intact. Native optimizers hide inactive public groups at explicit step.
2. **ODE precision boundary.** ODE was the only scheduler branch missing the historical storage-dtype round-trip. Its differentiable fp32 mean is now quantized through the rollout input dtype while retaining the graph. Replay diagnostics report ULPs in each original dtype.

Also moved random DMD/TDM ODE boundary selection off Python global RNG and onto the scheduler's epoch seed + shared random-step selector, reused by SDE selection. All ranks/components now choose the same reproducible boundary.

Verification: 1208 tests pass; 8-GPU ZeRO-2 DMD2 runs for two epochs at fake:gen 1:1 and 5:1; TDM runs for two epochs; and DMD2 5:1 passes with strict `replay_rtol=0`, `replay_atol=0` (bit-exact).

**Comment by Jayce-Ping** [comment]
## Replay/state root cause fixed

Pushed `a0f17ed` and `9cb97fc`.

The two-epoch replay failure was not checkpointing and not a tolerance issue. ULP diagnostics showed 16–512 fp16 ULP discrepancies. A pre-fake replay passed, and parameter probes then proved one fake ZeRO-2 step mutated 190/382 generator LoRA tensors (`max_abs=5.75e-06`) despite public generator grads being `None`.

Two fixes were required:

1. **ZeRO-2 role isolation.** Accelerate's DeepSpeed wrapper steps inside `backward()` and ZeRO iterates every internal `bit16_group`; native `grad is None` semantics do not apply. The role coordinator now filters DeepSpeed's per-group optimizer seam during backward, preserving inactive params, Adam moments, and step counters while leaving reduction/cleanup/all-gather intact. Native optimizers hide inactive public groups at explicit step.
2. **ODE precision boundary.** ODE was the only scheduler branch missing the historical storage-dtype round-trip. Its differentiable fp32 mean is now quantized through the rollout input dtype while retaining the graph. Replay diagnostics report ULPs in each original dtype.

Also moved random DMD/TDM ODE boundary selection off Python global RNG and onto the scheduler's epoch seed + shared random-step selector, reused by SDE selection. All ranks/components now choose the same reproducible boundary.

Verification: 1208 tests pass; 8-GPU ZeRO-2 DMD2 runs for two epochs at fake:gen 1:1 and 5:1; TDM runs for two epochs; and DMD2 5:1 passes with strict `replay_rtol=0`, `replay_atol=0` (bit-exact).

**Comment by Jayce-Ping** [comment]
## TDM official conditional-noise alignment

Pushed `297df4e`.

The scalar TDM fake/generator losses were already algebraically correct, but their inputs were not: Flow-Factory treated intermediate `x_{i+1}` as a clean prediction, applied ordinary forward noise, and passed unit importance weights. It now replays each stage velocity into `x0_hat_i`, derives the paired generator noise, conditionally renoises inside `[sigma_{i+1}, sigma_i]`, and applies the official mixed-vs-fresh Gaussian density ratio (clipped at configurable `train.tdm_importance_clip`, default 20) before SNR-capped fake DSM. Generator Pseudo-Huber/MSE and its detached normalizer remain unchanged.

`gen-ar` already implements the same equations in `tdm_utils.conditional_renoise_linear`; its self-test passes and required no core change.

Verification: 1210 unit tests pass; a 16-GPU, 2-node SD3.5 TDM DDP smoke completed two epochs successfully with strict replay tolerance while the existing DMD long run continued concurrently.

**Comment by Jayce-Ping** [comment]
## FSDP2 multi-role bug fixed

Pushed `e76d9aa`.

Root cause: FSDP2 replaces original Parameters with DTensor-backed Parameters during `prepare`, and activation checkpointing inserts `_checkpoint_wrapped_module` into qualified names. Prepared optimizer roles were already rebound, but `ComponentVariantRegistry` still owned pre-prepare identities. After a nested PEFT role context restored the generator adapter, registry trainability restoration wrote to stale tensors, leaving fake/surrogate live DTensors frozen (`fake: 382 params, 0 require grad`).

Fix: immediately after prepare, canonicalize transparent wrapper segments and exhaustively rebind every `(variant, component, parameter_name)` record against the prepared unwrapped bundle; rebuild identity ownership, validate route/name/shape uniqueness, then restore trainability on live DTensors before constructing role optimization.

8-GPU FSDP2 verification, all successful with 0 tracebacks:
- DMD2 LoRA
- TDM LoRA
- TDM-R1 LoRA (including slow-surrogate snapshot/trust region)
- DMD2 full
- TDM full

TDM-R1 full remains the pre-declared OOM exception. Full test suite: 1212 passed, 14 skipped.

File: tests/trainers/test_tdm_r1.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+from types import MethodType, SimpleNamespace
+from typing import Dict, Tuple
```
[the remark this answers is not in the document]
Comment (Jayce-Ping) [reply tests/trainers/test_tdm_r1.py]: Fixed in c0a3f8b by importing `List`; the focused TDM-R1 suite passes (27 tests).
